src/main/scala/is/hail/annotations/StagedRegionValueBuilder.scala

```diff
@@ -0,0 +1,174 @@
+package is.hail.annotations
+
+import is.hail.asm4s.{AsmFunction2, Code, FunctionBuilder}
+import is.hail.asm4s.Code._
+import is.hail.asm4s._
+import is.hail.expr._
+import is.hail.utils._
+import org.objectweb.asm.tree.{AbstractInsnNode, IincInsnNode}
+
+import scala.collection.generic.Growable
+import scala.reflect.ClassTag
+
+class StagedRegionValueBuilder[T] private (val fb: FunctionBuilder[AsmFunction2[T, MemoryBuffer, Long]], val rowType: Type, var region: Code[MemoryBuffer], val pOffset: Code[Long])(implicit tti: TypeInfo[T]) {
+
+  private def this(fb: FunctionBuilder[AsmFunction2[T, MemoryBuffer, Long]], rowType: Type, parent: StagedRegionValueBuilder[T])(implicit tti: TypeInfo[T]) = {
+    this(fb, rowType, parent.region, parent.currentOffset)
+  }
+  def this(fb: FunctionBuilder[AsmFunction2[T, MemoryBuffer, Long]], rowType: Type)(implicit tti: TypeInfo[T]) = {
+    this(fb, rowType, fb.getArg[MemoryBuffer](2), null)
+  }
+
+  val input: LocalRef[T] = fb.getArg[T](1)
+  var staticIdx: Int = 0
+  var idx: LocalRef[Int] = _
+  var elementsOffset: LocalRef[Long] = _
+
+  var elementsOffsetArray: LocalRef[Array[Long]] = _
```

Comment:
If these are not intended to be accessed outside of this class, let's make them `private`. Also, our code style prefers no space between the list of fields unless it contributes significantly to understanding.

Reply:
I think being able to look at idx can be pretty handy because if you're iterating through an array anyways it's nice to not have to keep your own iterator, but you're right that being able to modify the stored value is a bit dicey. I'm going to make it private and then add an acceessor method?
```
def arrayIdx: Code[Int] = idx.load()
```

Reply:
Sounds good.